### PR TITLE
[NL] Remove vector<unique_ptr> from DOF table.

### PR DESCRIPTION
**Pull request by endJunction**
The main change is in the LocalToGlobalIndexMap, all other changes are "just" fixing the compilation errors. In the processes not all of the unique_ptr's are removed, since somebody has to be responsible for the MeshSubset.

### CODE REVIEW

File: NumLib/DOF/LocalToGlobalIndexMap.cpp [hunk 1]
```diff
@@ -248,13 +246,13 @@ LocalToGlobalIndexMap* LocalToGlobalIndexMap::deriveBoundaryConstrainedMap(
...
-        _mesh_component_map.getSubset(global_component_ids, *mesh_subsets);
+        _mesh_component_map.getSubset(global_component_ids, mesh_subsets);
 
     // Create copies of the mesh_subsets for each of the global components.
     // The last component is moved after the for-loop.
-    std::vector<std::unique_ptr<MeshLib::MeshSubsets>> all_mesh_subsets;
+    std::vector<MeshLib::MeshSubsets> all_mesh_subsets;
     for (int i = 0; i < static_cast<int>(global_component_ids.size()) - 1; ++i)
-        all_mesh_subsets.emplace_back(new MeshLib::MeshSubsets{*mesh_subsets});
+        all_mesh_subsets.emplace_back(MeshLib::MeshSubsets{mesh_subsets});
```
Comment (chleh) [inline NumLib/DOF/LocalToGlobalIndexMap.cpp]: `MeshLib::MeshSubsets` can be omitted, I guess.

File: ProcessLib/CalculateSurfaceFlux/CalculateSurfaceFlux.cpp [hunk 2]
```diff
@@ -29,14 +29,11 @@ CalculateSurfaceFlux::CalculateSurfaceFlux(
...
-    std::vector<std::unique_ptr<MeshLib::MeshSubsets>> all_mesh_subsets;
+    std::vector<MeshLib::MeshSubsets> all_mesh_subsets;
     std::generate_n(
         std::back_inserter(all_mesh_subsets),
         bulk_property_number_of_components,
-        [&]() {
-            return std::unique_ptr<MeshLib::MeshSubsets>{
-                new MeshLib::MeshSubsets{mesh_subset_all_nodes.get()}};
-        });
+        [&]() { return MeshLib::MeshSubsets{mesh_subset_all_nodes.get()}; });
```
Comment (chleh) [inline ProcessLib/CalculateSurfaceFlux/CalculateSurfaceFlux.cpp]: `MeshLib::MeshSubsets` not needed.

File: ProcessLib/HydroMechanics/HydroMechanicsProcess.h [hunk 3]
```diff
@@ -71,17 +71,16 @@ class HydroMechanicsProcess final : public Process
         // Collect the mesh subsets in a vector.
 
         // For pressure, which is the first
-        std::vector<std::unique_ptr<MeshLib::MeshSubsets>> all_mesh_subsets;
-        all_mesh_subsets.push_back(std::unique_ptr<MeshLib::MeshSubsets>{
-            new MeshLib::MeshSubsets{_mesh_subset_base_nodes.get()}});
+        std::vector<MeshLib::MeshSubsets> all_mesh_subsets;
+        all_mesh_subsets.push_back(
+            MeshLib::MeshSubsets{_mesh_subset_base_nodes.get()});
```
Comment (chleh) [inline ProcessLib/HydroMechanics/HydroMechanicsProcess.h]: Maybe `all_mesh_subsets.emplace_back(_mesh_subset_base_nodes.get())` ?

File: ProcessLib/LIE/HydroMechanics/HydroMechanicsProcess.cpp [hunk 4]
```diff
@@ -143,13 +143,13 @@ void HydroMechanicsProcess<GlobalDim>::constructDofTable()
...
-    std::vector<std::unique_ptr<MeshLib::MeshSubsets>> all_mesh_subsets;
+    std::vector<MeshLib::MeshSubsets> all_mesh_subsets;
     std::vector<unsigned> vec_n_components;
     std::vector<std::vector<MeshLib::Element*> const*> vec_var_elements;
     // pressure
     vec_n_components.push_back(1);
-    all_mesh_subsets.push_back(std::unique_ptr<MeshLib::MeshSubsets>{
-        new MeshLib::MeshSubsets{_mesh_subset_nodes_p.get()}});
+    all_mesh_subsets.push_back(
+        MeshLib::MeshSubsets{_mesh_subset_nodes_p.get()});
```
Comment (chleh) [inline ProcessLib/LIE/HydroMechanics/HydroMechanicsProcess.cpp]: likewise.

File: ProcessLib/LIE/SmallDeformation/SmallDeformationProcess.cpp [hunk 5]
```diff
@@ -183,10 +175,9 @@ void SmallDeformationProcess<DisplacementDim>::initializeConcreteProcess(
 
     // TODO move the two data members somewhere else.
     // for extrapolation of secondary variables
-    std::vector<std::unique_ptr<MeshLib::MeshSubsets>>
-        all_mesh_subsets_single_component;
+    std::vector<MeshLib::MeshSubsets> all_mesh_subsets_single_component;
     all_mesh_subsets_single_component.emplace_back(
-        new MeshLib::MeshSubsets(_mesh_subset_all_nodes.get()));
+        MeshLib::MeshSubsets{_mesh_subset_all_nodes.get()});
```
Comment (chleh) [inline ProcessLib/LIE/SmallDeformation/SmallDeformationProcess.cpp]: likewise.

File: ProcessLib/Process.cpp [hunk 6]
```diff
@@ -186,10 +185,9 @@ void Process::initializeExtrapolator()
     else
     {
         // Otherwise construct a new DOF table.
-        std::vector<std::unique_ptr<MeshLib::MeshSubsets>>
-            all_mesh_subsets_single_component;
+        std::vector<MeshLib::MeshSubsets> all_mesh_subsets_single_component;
         all_mesh_subsets_single_component.emplace_back(
-            new MeshLib::MeshSubsets(_mesh_subset_all_nodes.get()));
+            MeshLib::MeshSubsets(_mesh_subset_all_nodes.get()));
```
Comment (chleh) [inline ProcessLib/Process.cpp]: similar.

File: ProcessLib/CalculateSurfaceFlux/CalculateSurfaceFlux.cpp (hunk 2 above)
Comment (endJunction) [inline ProcessLib/CalculateSurfaceFlux/CalculateSurfaceFlux.cpp]: You are right, it cannot be done here:
```
error: cannot deduce lambda return type from initializer list
```

**Pull request merged**